CMakeLists.txt

```diff
@@ -161,33 +179,33 @@ if(BUILD_JS)
     file(GLOB common-fonts LIST_DIRECTORIES false ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common/fonts/*)
 
     # Builds the Common Sys Lib
-    file(GLOB common_lib_sources ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common-lib/src/*)
-    add_custom_command(
-        OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common-lib/dist/common-lib.js
-        DEPENDS ${common_lib_sources}
-        COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common-lib && yarn --mutex network && yarn build
-    )
-    add_custom_target(CommonSysLib_js
-        DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common-lib/dist/common-lib.js
-    )
+    add_js_target(${COMMON_DIR} common-lib.js)
 
     foreach(ENTRY ${JS_DIRS})
-        string(REGEX REPLACE "^([^:]+):([^:]+)$" \\1 PATH ${ENTRY})
-        string(REGEX REPLACE "^([^:]+):([^:]+)$" \\2 DEP ${ENTRY})
-        file(GLOB_RECURSE sources ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/src/*)
-        add_custom_command(
-            OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/dist/index.html
-            DEPENDS ${sources} ${common} ${common-fonts} CommonSysLib_js
-            COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR}/${PATH} && yarn --mutex network && yarn build
-        )
-        add_custom_target(
-            ${DEP}
-            DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/dist/index.html
-        )
+        add_js_target(${ENTRY} index.html ${common} ${common-fonts} CommonSysLib_js)
     endforeach()
-endif()
 
-if(BUILD_JS)
+    set(byproducts ${CMAKE_CURRENT_SOURCE_DIR}/services/user/AdminSys/ui/wasm-psibase)
+
+    add_js_target(${ADMIN_DIR} index.html
+        ${byproducts}/wasm-psibase_bg.js
+        ${byproducts}/wasm-psibase_bg.wasm
+        ${byproducts}/wasm-psibase_bg.wasm.d.ts
+        ${byproducts}/wasm-psibase.d.ts
+        ${byproducts}/wasm-psibase.js
+    )
+
+    add_custom_target(wasm-psibase ALL
```

Comment:
`wasm-pack` always updates the timestamp on its output files. I had assumed that it was checking the source timestamps because it uses cargo, but that only applies to the wasm build step. Other steps run unconditionally. This causes AdminSys to rebuild every time, which is what I was hoping to avoid.

These infinite rebuilds can be fixed by moving the `wasm-pack` command into an `add_custom_command` associated with the `wasm-psibase` target. In order to ensure that it rebuilds when the sources are modified, add `BUILD_BYPRODUCTS` pointing at the `psibase` executable to the `rust` external project and make the custom command depend on `psibase`. This is a somewhat hacky, as it actually depends on the *sources* of `psibase`, but it should work well enough in practice.

Reply:
This setup still seems to build `wasm-psibase` on every build, presumably because `rust` is also building on every build.
It's not building everything, but it's doing something... I see this on every build, no matter if I change sources:
```
[  9%] Performing build step for 'rust'
   Compiling psibase v0.2.0 (/root/psibase/rust/psibase)
    Finished release [optimized] target(s) in 34.57s
[  9%] No install step for 'rust'
[  9%] Completed 'rust'
[ 10%] Built target rust
[ 10%] Generating ../services/user/AdminSys/ui/wasm-psibase/wasm-psibase_bg.js, ../services/user/AdminSys/ui/wasm-psibase/wasm-psibase_bg.wasm, ../services/user/AdminSys/ui/wasm-psibase/wasm-psibase_bg.wasm.d.ts, ../services/user/AdminSys/ui/wasm-psibase/wasm-psibase.d.ts, ../services/user/AdminSys/ui/wasm-psibase/wasm-psibase.js
[INFO]: 🎯  Checking for the Wasm target...
[INFO]: 🌀  Compiling to Wasm...
   Compiling psibase v0.2.0 (/root/psibase/rust/psibase)
    Finished release [optimized] target(s) in 22.57s
```